### PR TITLE
Feature/leader config

### PR DESCRIPTION
**Pull request by tpetracca**
Per our internal thread re: Atlas leader blocks.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/593)

### CODE REVIEW

**Comment by tpetracca** [comment]
So in order to get this into a good place we'd have to fix the atlasdb-cli testing to actually spin up an external leader that the cli can point to.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java [hunk 1]
```diff
@@ -22,16 +22,21 @@
...
+import com.google.common.base.Optional;
 import com.google.common.base.Preconditions;
 
 @JsonDeserialize(as = ImmutableLeaderConfig.class)
 @JsonSerialize(as = ImmutableLeaderConfig.class)
 @Value.Immutable
 public abstract class LeaderConfig {
 
-    public abstract int quorumSize();
+    @JsonIgnore
```
Comment (uschi2000) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java]: check out derived attributes: http://immutables.github.io/immutable.html#derived-attributes

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java [hunk 2]
```diff
@@ -65,8 +74,10 @@ public long leaderPingResponseWaitMs() {
 
     @Value.Check
     protected final void check() {
-        Preconditions.checkArgument(leaders().contains(localServer()),
-                "The localServer '%s' must included in the leader entries %s.", localServer(), leaders());
+        Preconditions.checkArgument(!isLeader() || localServer().isPresent(),
+                "Local server isn't set even though we are set to be a leader");
+        Preconditions.checkArgument(leaders().size() == leaderCount(),
+                "Number of leaders {} does not equal the expected leader count {}", leaders().size(), leaderCount());
```
Comment (rjullman) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java]: I would also make leaderCount derived if you want this constraint always.

Comment (tpetracca) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java]: Followed up offline with Bobby about how I want this for catching misconfigurations in leader lists.

**Comment by gsheasby** [comment]
Just for tracking: this would require a config change in internal products that depend on us. If we go through with this, then we should make sure that the error message is *really* clear.

**Comment by joelea** [comment]
I think the fix we want to make for this kind of effect is #645 - this should be happening in the next month or so and will require more config changes. My feeling is that we should skip this in preference for that change.

**Comment by joelea** [comment]
This is more important than originally thought - we need to do the appropriate communication with dependent products and get this merged to make sure everyone has an exposed timestamp and lock server.

Without this, no one will be able to run sweep when it becomes necessary.

**Comment by joelea** [comment]
Closing this in favour of #708

**Pull request closed without merging**